### PR TITLE
UNIT Tests using LIT framework

### PR DESCRIPTION
**Pull request by shubhamnarlawar**
Added nballerina testcases inside ./llvm/test/CodeGen directory.

### CODE REVIEW

**Comment by shubhamnarlawar** [comment]
From #28 suggestions, following things has been fixed - 
1. Renamed `NBallerian` to `NBallerina`.
2. Renamed generic testcases names such as wso2testcase.bal, etc
3. Removed `TESTBUILD_PATH` flag.
4. Removed "--dump-bir" flag from the script.

Issues remaining - 
1. As pointed out by @uthaiyashankar, in simpleMain.bal,
RETVAL - 112 (instead of 128) on Mac machine. This issue will go once rust routines PR is merged. We use rust print routines instead of `echo $? of Linux`.

2. Finding correct target triple generation API. @ayonam is assiting us on this issue.

**Comment by shubhamnarlawar** [comment]
Steps to run nballerina unit tests using LIT framework - 
1. Set the paths in set_env file present in BIR2llvmIR. (BIR2llvmIR will be renamed to "compiler" once all the existing PR are merged).
2. source set_env
3. `./build/llvm/bin/llvm-lit ./llvm-project-master/llvm/test/NBallerina/`

I will put above comments in the README

File: BIR2llvmIR/llvmBinaryOpInsn.cpp [hunk 1]
```diff
@@ -21,6 +21,8 @@ void BinaryOpInsn::translate(LLVMModuleRef &modRef) {
     lhsName = getLhsOperand()->name();
     lhstmpName = lhsName + "_temp";
     lhsRef = funcObj->getLocalVarRefUsingId(lhsName);
+    if(!lhsRef)
```
Comment (warunalakshitha) [inline BIR2llvmIR/llvmBinaryOpInsn.cpp:24]:
```suggestion
    if (!lhsRef)
```

File: BIR2llvmIR/llvmUnaryOpInsn.cpp [hunk 2]
```diff
@@ -22,6 +22,8 @@ void UnaryOpInsn::translate(LLVMModuleRef &modRef) {
     builder = funcObj->getLLVMBuilder();
     lhstmpName = lhsName + "_temp";
     lhsRef = funcObj->getLocalVarRefUsingId(lhsName);
+    if(!lhsRef)
```
Comment (warunalakshitha) [inline BIR2llvmIR/llvmUnaryOpInsn.cpp:25]:
```suggestion
    if (!lhsRef)
```

File: llvm-project-master/llvm/test/NBallerina/allBinaryOperations.bal [hunk 3]
```diff
@@ -0,0 +1,30 @@
+// RUN: JAVA_HOME=%JAVA_HOME WORK_PATH=%WORK_PATH %WORK_PATH/testRunScript.sh %s -o - | FileCheck %s
+
+int _bal_result = 0;
+public function main() {
+   int a = 10;
```
Comment (warunalakshitha) [inline llvm-project-master/llvm/test/NBallerina/allBinaryOperations.bal:5]: Seems minor alignment issue.

File: llvm-project-master/llvm/test/NBallerina/allop.bal [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    if (a >= b) {
+      f = false;
+    }
+    else {
+      f = true;
+    }
+    while (i < 10) {
+      i = i + 1;
+    }
+    _bal_result = a;
```
Comment (warunalakshitha) [inline llvm-project-master/llvm/test/NBallerina/allop.bal:40]: Don't we need to check other values(b,c,d etc) or is this test covers only codegen part? . We can increment the bal_result for each condition and check the final value.

File: llvm-project-master/llvm/test/NBallerina/allBinaryOperations.bal (hunk 3 above)
Comment (shubhamnarlawar) [reply llvm-project-master/llvm/test/NBallerina/allBinaryOperations.bal:5]: Do you want me to run clang-format on .bal files also

**Comment by ayonam** [comment]
Shubham,
    For .bal files, I think you would have a
        different coding standard.  Either you will have to follow that
        or you will have to write a small tool to do the formatting for
        you.  I am not sure if such a tool already exists.
       
    
      
      
    
    On 01/02/21 14:26, shubhamnarlawar
      wrote:
    
    
      
      @shubhamnarlawar commented on this pull request.
      
      In llvm-project-master/llvm/test/NBallerina/allBinaryOperations.bal:
      > @@ -0,0 +1,30 @@
+// RUN: JAVA_HOME=%JAVA_HOME WORK_PATH=%WORK_PATH %WORK_PATH/testRunScript.sh %s -o - | FileCheck %s
+
+int _bal_result = 0;
+public function main() {
+   int a = 10;

      Do you want me to run clang-format on .bal files also
      —
        You are receiving this because you were mentioned.
        Reply to this email directly, view it on GitHub, or unsubscribe.
      [
{
"@context": "http://schema.org",
"@type": "EmailMessage",
"potentialAction": {
"@type": "ViewAction",
"target": "https://github.com/ballerina-platform/nballerina/pull/47#discussion_r567654503",
"url": "https://github.com/ballerina-platform/nballerina/pull/47#discussion_r567654503",
"name": "View Pull Request"
},
"description": "View this Pull Request on GitHub",
"publisher": {
"@type": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

**Comment by shubhamnarlawar** [comment]
> Shubham, For .bal files, I think you would have a different coding standard.  Either you will have to follow that or you will have to write a small tool to do the formatting for you.  I am not sure if such a tool already exists. On 01/02/21 14:26, shubhamnarlawar wrote: @shubhamnarlawar commented on this pull request. In llvm-project-master/llvm/test/NBallerina/allBinaryOperations.bal: @@ -0,0 +1,30 @@ +// RUN: JAVA_HOME=%JAVA_HOME WORK_PATH=%WORK_PATH %WORK_PATH/testRunScript.sh %s -o - | FileCheck %s + +int _bal_result = 0; +public function main() { + int a = 10; Do you want me to run clang-format on .bal files also — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub, or unsubscribe. [ { "@context": "http://schema.org", "@type": "EmailMessage", "potentialAction": { "@type": "ViewAction", "target": "[#47 (comment)](https://github.com/ballerina-platform/nballerina/pull/47#discussion_r567654503)", "url": "[#47 (comment)](https://github.com/ballerina-platform/nballerina/pull/47#discussion_r567654503)", "name": "View Pull Request" }, "description": "View this Pull Request on GitHub", "publisher": { "@type": "Organization", "name": "GitHub", "url": "https://github.com" } } ]

Got it.

@warunalakshitha - Is there such tool in ballerina community that does formatting of .bal files

If not then, we need to manually format all the testcases!

File: llvm-project-master/llvm/test/NBallerina/allop.bal (hunk 4 above)
Comment (shubhamnarlawar) [reply llvm-project-master/llvm/test/NBallerina/allop.bal:40]: Yes. I think incrementing bal_result for each conditional loop seems a good suggestion. I will update the testcase.

**Comment by ruvi-d** [comment]
@shubhamnarlawar : try the `ballerina format` command as described here: https://ballerina.io/swan-lake/learn/using-the-cli-tools/#core-commands

**Comment by ayonam** [comment]
Shubham,
    Aren't we using the print routines as yet?  Or is it scheduled
      for a later PR?
       
    
    
    
    On 01/02/21 14:56, shubhamnarlawar
      wrote:
    
    
      
      @shubhamnarlawar commented on this pull request.
      
      In llvm-project-master/llvm/test/NBallerina/allop.bal:
      > +    if (a > b) {
+      f = true;
+    }
+    if ( a <= b) {
+      e = true;
+    }
+    if (a >= b) {
+      f = false;
+    }
+    else {
+      f = true;
+    }
+    while (i < 10) {
+      i = i + 1;
+    }
+    _bal_result = a;

      Yes. I think incrementing bal_result for each conditional loop
        seems a good suggestion. I will update the testcase.
      —
        You are receiving this because you were mentioned.
        Reply to this email directly, view it on GitHub, or unsubscribe.
      [
{
"@context": "http://schema.org",
"@type": "EmailMessage",
"potentialAction": {
"@type": "ViewAction",
"target": "https://github.com/ballerina-platform/nballerina/pull/47#discussion_r567673174",
"url": "https://github.com/ballerina-platform/nballerina/pull/47#discussion_r567673174",
"name": "View Pull Request"
},
"description": "View this Pull Request on GitHub",
"publisher": {
"@type": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

**Comment by shubhamnarlawar** [comment]
> @shubhamnarlawar : try the `ballerina format` command as described here: https://ballerina.io/swan-lake/learn/using-the-cli-tools/#core-commands

Done. This is helpful. Thanks for sharing it!

**Comment by shubhamnarlawar** [comment]
> Shubham, Aren't we using the print routines as yet?  Or is it scheduled for a later PR? On 01/02/21 14:56, shubhamnarlawar wrote: @shubhamnarlawar commented on this pull request. In llvm-project-master/llvm/test/NBallerina/allop.bal: + if (a > b) { + f = true; + } + if ( a <= b) { + e = true; + } + if (a >= b) { + f = false; + } + else { + f = true; + } + while (i < 10) { + i = i + 1; + } + _bal_result = a; Yes. I think incrementing bal_result for each conditional loop seems a good suggestion. I will update the testcase. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub, or unsubscribe. [ { "@context": "http://schema.org", "@type": "EmailMessage", "potentialAction": { "@type": "ViewAction", "target": "[#47 (comment)](https://github.com/ballerina-platform/nballerina/pull/47#discussion_r567673174)", "url": "[#47 (comment)](https://github.com/ballerina-platform/nballerina/pull/47#discussion_r567673174)", "name": "View Pull Request" }, "description": "View this Pull Request on GitHub", "publisher": { "@type": "Organization", "name": "GitHub", "url": "https://github.com" } } ]

Not yet. Rust routines are present in #41 which is still under review.
We will schedule using rust print routines in LIT unit tests in separate PR.

File: llvm-project-master/llvm/test/NBallerina/allop.bal (hunk 4 above)
Comment (ramakotareddy) [reply llvm-project-master/llvm/test/NBallerina/allop.bal:40]: Yes. I have updated the test case as per Waruna suggestion.

**Review by ruvi-d — APPROVED** [review]

**Review by uthaiyashankar — APPROVED** [review]

**Pull request merged**